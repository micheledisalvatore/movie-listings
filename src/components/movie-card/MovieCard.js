import React from 'react';
import PropTypes from 'prop-types';
import { isEmpty } from 'ramda';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Typography from '@material-ui/core/Typography';
import Chip from '@material-ui/core/Chip';

export const MovieCard = ({ classes, image, title, genreIds, genres }) => (
  <Card className={classes.card}>
    <div className={classes.details}>
      <CardMedia
        component="img"
        alt="Contemplative Reptile"
        className={classes.media}
        image={`https://image.tmdb.org/t/p/w500/${image}`}
        title={title}
      />
      <CardContent>
        <Typography gutterBottom variant="h5" component="h2">
          {title}
        </Typography>
        <div>
        {!isEmpty(genres) && genreIds.map(genreId => {
          const { name } = genres[genreId];
          return (
            <Chip
              key={genreId}
              label={name}
              className={classes.chip}
            />
          )}
        )}
        </div>
      </CardContent>
    </div>
  </Card>
);

MovieCard.propTypes = {
  classes: PropTypes.object.isRequired,
  genreIds: PropTypes.arrayOf(PropTypes.number),
  genres: PropTypes.shape({
    name: PropTypes.string,
  }),
  image: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
};

MovieCard.defaultProps = {
  genreIds: [],
  genres: {},
}

export const mapStateToProps = ({ genres: { list }}) => ({
  genres: list || {},
})