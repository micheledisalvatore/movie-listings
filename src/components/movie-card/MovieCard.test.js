/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import CardMedia from '@material-ui/core/CardMedia';
import Chip from '@material-ui/core/Chip';

import { MovieCard, mapStateToProps } from './MovieCard';

describe('Given a MovieCard component', () => {
  let component;
  let wrapper;
  const requiredProps = {
    classes: {},
    image: 'fooImageUrl',
    title: 'fooTitle',
  };

  describe('when it is rendered', () => {
    beforeEach(() => {
      component = <MovieCard {...requiredProps} />;
      wrapper = shallow(component);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the passed image', () => {
      expect(wrapper.find(CardMedia)).toHaveProp('image', 'https://image.tmdb.org/t/p/w500/fooImageUrl');
    });

    it('should render the image with the title as placeholder', () => {
      expect(wrapper.find(CardMedia)).toHaveProp('title', 'fooTitle');
    });

    it('should NOT render the genre chips', () => {
      expect(wrapper.find(Chip)).not.toExist();
    });

    describe('and the list of genres is populated', () => {
      beforeEach(() => {
        wrapper.setProps({
          genres: {
            0: { name: 'foo' },
            1: { name: 'bar' },
          },
        });
      });

      it('should NOT render the genre chips', () => {
        expect(wrapper.find(Chip)).not.toExist();
      });

      describe('and the list of genreIds is populated', () => {
        beforeEach(() => {
          wrapper.setProps({
            genreIds: [0],
          });
        });

        it('should render one genre chip', () => {
          expect(wrapper.find(Chip)).toHaveLength(1);
        });

        it('should render the chip with the right name', () => {
          expect(wrapper.find(Chip)).toHaveProp('label', 'foo');
        });
      });
    });
  });
});

describe('Given a mapStateToProps function', () => {
  describe('when it is called', () => {
    it('should return an object with the passed values', () => {
      expect(mapStateToProps({
        genres: {
          list: {
            1: {
              id: 1,
              name: 'foo',
            },
            3: {
              id: 3,
              name: 'bar',
            },
          },
        },
      })).toEqual({
        genres: {
          1: {
            id: 1,
            name: 'foo',
          },
          3: {
            id: 3,
            name: 'bar',
          },
        },
      });
    });
  });
});
