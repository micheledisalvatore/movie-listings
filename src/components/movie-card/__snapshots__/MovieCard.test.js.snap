// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a MovieCard component when it is rendered should match the snapshot 1`] = `
<WithStyles(Card)>
  <div>
    <WithStyles(CardMedia)
      alt="Contemplative Reptile"
      component="img"
      image="https://image.tmdb.org/t/p/w500/fooImageUrl"
      title="fooTitle"
    />
    <WithStyles(CardContent)>
      <WithStyles(Typography)
        component="h2"
        gutterBottom={true}
        variant="h5"
      >
        fooTitle
      </WithStyles(Typography)>
      <div />
    </WithStyles(CardContent)>
  </div>
</WithStyles(Card)>
`;
